### PR TITLE
Refactor walt process to support Picard metrics

### PR DESCRIPTION
**Pull request by AnzeLovse**
## Checklist

* [ ] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [ ] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [ ] All inputs are used in process.
* [ ] All output fields have their ``re-save`` calls.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of tool should be self explanatory with links to tool's homepage
  and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required: false`` is not needed.
* Add descriptive error and warning messages.
* Place a ``re-checkrc`` call after each process step (call of a tool, a
  Bash command, a Python script or an R script).
* Add ``re-progress`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Changes to Docker images need to be accepted in a separate pull request.
  Before the new images can be used for automated testing of your PR, they
  must be manually tagged and pushed to Docker Hub.
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files with spaces:
  * Genomes: ``genome species.fa.gz``
  * Annotations: ``annotation species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Output and specific input files should include the name of the tested tool.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if length of some list is 0, if string is given where one expects
  integer, if file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

File: resolwe_bio/processes/WGBS/walt.yml [hunk 1]
```diff
@@ -137,25 +163,13 @@
...
-        if [ -s "${NAME}.mr_1_unmapped" ]; then
-          gzip "${NAME}.mr_1_unmapped"
-          re-save-file unmapped_f "${NAME}.mr_1_unmapped.gz"
-        fi
-        if [ -s "${NAME}.mr_2_unmapped" ]; then
-          gzip "${NAME}.mr_2_unmapped"
-          re-save-file unmapped_r "${NAME}.mr_2_unmapped.gz"
-        fi
-      {% endif %}
+      gzip "${NAME}_unmapped.bam"
```
Comment (romunov) [inline resolwe_bio/processes/WGBS/walt.yml]: @jkokosar is this the fastest compression tool we use?

Comment (AnzeLovse) [reply resolwe_bio/processes/WGBS/walt.yml]: Thanks for reminding me, should I use pigz for compression?

File: resolwe_bio/processes/WGBS/methcounts.yml [hunk 2]
```diff
@@ -11,7 +11,7 @@
       docker:
         image: resolwebio/wgbs:1.0.0
```
Comment (jkokosar) [inline resolwe_bio/processes/WGBS/methcounts.yml]: Please bump the process to the latest version of the `resolwebio/wgbs` to minimise the usage of different versions of Docker images.

File: resolwe_bio/processes/WGBS/walt.yml [hunk 3]
```diff
@@ -45,19 +45,26 @@
         Sets the number of reads to mapping in each loop. Larger number results in program taking
         more memory. This is especially evident in paired-end mapping.
   output:
-    - name: mr
+    - name: bam
       label: Alignment file
```
Comment (jkokosar) [inline resolwe_bio/processes/WGBS/walt.yml]: Update the labels for the `bam` and `mr` so that it will be possible to differentiate between the the output fields in the UI. One option is to use `Alignment file (BAM)` and `Alignment file (MR)`.

File: resolwe_bio/processes/WGBS/walt.yml [hunk 4]
```diff
@@ -45,19 +45,26 @@
...
+    - name: bigwig
+      label: BigWig file
       type: basic:file
       required: false
-    - name: unmapped_r
-      label: Unmapped reads (mate 2)
+    - name: mr
+      label: Alignment file
+      type: basic:file
+      description: Position sorted alignment
```
Comment (jkokosar) [inline resolwe_bio/processes/WGBS/walt.yml]: Rewrite all the descriptions into format:
```
description: |
 Description text...
```

File: resolwe_bio/processes/WGBS/walt.yml [hunk 5]
```diff
@@ -113,12 +120,36 @@
...
+        samtools view "${NAME}.bam" -f 0x4 -o "${NAME}_unmapped.bam"
+      {% else %}
+        samtools view "${NAME}.bam" -f 0x2 -o "${NAME}_mapped.bam"
+        samtools view "${NAME}.bam" -F 0x2 -o "${NAME}_unmapped.bam"
+      {% endif %}
+
+      samtools sort \
+            -@ {{ requirements.resources.cores }} \
+            -m {{ (requirements.resources.memory * 1024 * 1024 / requirements.resources.cores * 0.8) | int }} \
+            -o "${NAME}_sorted.bam"\
```
Comment (jkokosar) [inline resolwe_bio/processes/WGBS/walt.yml]: Missing space between `.bam"\`?

File: resolwe_bio/processes/WGBS/walt.yml (hunk 1 above)
Comment (jkokosar) [reply resolwe_bio/processes/WGBS/walt.yml]: It's unnecessary to compress the `.bam`, since this is already a compressed binary format of the `.sam` data type.

File: resolwe_bio/processes/WGBS/walt.yml [hunk 6]
```diff
@@ -137,25 +168,13 @@
       gzip "${NAME}.mr"
```
Comment (jkokosar) [inline resolwe_bio/processes/WGBS/walt.yml]: Use the multithreaded `pigz` instead of `gzip`.

File: resolwe_bio/processes/WGBS/walt.yml [hunk 7]
```diff
@@ -137,25 +168,13 @@
...
-          gzip "${NAME}.mr_1_unmapped"
-          re-save-file unmapped_f "${NAME}.mr_1_unmapped.gz"
-        fi
-        if [ -s "${NAME}.mr_2_unmapped" ]; then
-          gzip "${NAME}.mr_2_unmapped"
-          re-save-file unmapped_r "${NAME}.mr_2_unmapped.gz"
-        fi
-      {% endif %}
+      gzip "${NAME}_unmapped.bam"
+      re-save-file unmapped "${NAME}_unmapped.bam.gz"
```
Comment (jkokosar) [inline resolwe_bio/processes/WGBS/walt.yml]: Don't save the "compressed" `.bam`

File: resolwe_bio/processes/workflows/wgbs.yml [hunk 8]
```diff
@@ -61,6 +61,180 @@
...
+            - label: Unspecified
+              value: 'null'
+            - label: FR
+              value: FR
+            - label: RF
+              value: RF
+            - label: TANDEM
+              value: TANDEM
+          default: 'null'
+    - name: insert
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/wgbs.yml]: This block of inputs is not used, as `InsertSizeMetrics` is not used in the workflow.

File: resolwe_bio/processes/workflows/wgbs.yml [hunk 9]
```diff
@@ -83,3 +257,47 @@
...
+          mismatch_rate: '{{ input.rrbs_metrics.mismatch_rate }}'
+          validation_stringency: '{{ input.rrbs_metrics.validation_stringency }}'
+          assume_sorted: true
+      - id: multiqc
+        run: multiqc
+        input:
+          data: '{{ [
+            input.reads,
+            steps.wgs_metrics,
+            steps.rrbs_metrics,
```
Comment (jkokosar) [inline resolwe_bio/processes/workflows/wgbs.yml]: `steps.summary` is missing. Is it not supported?
